src/gs-background-tile.c

```diff
@@ -0,0 +1,303 @@
+
+	update_tile_colors_bg (tile);
+	g_object_unref (tile);
+
+	return G_SOURCE_REMOVE;
+}
+
+static void
+app_key_colors_changed (GsApp *app,
+			GParamSpec *pspec,
+			GsBackgroundTile *tile)
+{
+	/* we call the function in idle because the state-change
+	 * notification could be coming from a different thread */
+	g_idle_add (app_key_colors_changed_idle, g_object_ref (tile));
+}
+
+static void
+update_tile_price (GsBackgroundTile *tile)
+{
+	const GsPrice *price = NULL;
+	g_autofree gchar *price_str = NULL;
+
+	if (gs_app_is_installed (tile->app)) {
+		gtk_widget_hide (tile->price);
+		gtk_widget_show (tile->installed_icon);
+		return;
+	}
+
+	price = gs_app_get_price (tile->app);
+	if (price == NULL || (gs_price_get_amount (price) == (gdouble) 0)) {
+		/* TRANSLATORS: This is the price of the app, as in gratis */
+		price_str = g_utf8_strup (_("Free"), -1);
+		gtk_label_set_label (GTK_LABEL (tile->price), price_str);
+	} else {
+		price_str = gs_price_to_string (price);
+		gtk_label_set_label (GTK_LABEL (tile->price), price_str);
+	}
+
+	gtk_widget_show (tile->price);
+	gtk_widget_hide (tile->installed_icon);
+}
+
+static void
+update_tile_info (GsBackgroundTile *tile)
+{
+	g_autofree gchar *name = NULL;
+	AtkObject *accessible = gtk_widget_get_accessible (GTK_WIDGET (tile));
+	const gchar *css = NULL;
+	const gchar *summary = NULL;
+
+	if (gs_app_is_installed (tile->app)) {
+		/* TRANSLATORS: This is the name and state of an app for the ATK object */
+		name = g_strdup_printf (_("%s (Installed)"), gs_app_get_name (tile->app));
+	} else if (gs_app_get_state (tile->app) == AS_APP_STATE_AVAILABLE) {
+		name = g_strdup (gs_app_get_name (tile->app));
+	}
+
+	summary = gs_app_get_summary (tile->app);
+
+	if (GTK_IS_ACCESSIBLE (accessible)) {
+		atk_object_set_name (accessible, name);
+		atk_object_set_description (accessible, summary);
+	}
+
+	gtk_label_set_label (GTK_LABEL (tile->name_label), gs_app_get_name (tile->app));
+	gtk_label_set_label (GTK_LABEL (tile->summary_label), summary);
+
+	css = gs_app_get_metadata_item (tile->app, "GnomeSoftware::BackgroundTile-css");
+	if (css != NULL)
+		gs_utils_widget_set_css (GTK_WIDGET (tile->image_box), css);
+
+	update_tile_price (tile);
+}
+
+static gboolean
+app_state_changed_idle (gpointer user_data)
+{
+	GsBackgroundTile *tile = GS_BACKGROUND_TILE (user_data);
+
+	update_tile_info (tile);
+	g_object_unref (tile);
+
+	return G_SOURCE_REMOVE;
+}
+
+static void
+app_state_changed (GsApp *app,
+		   GParamSpec *pspec,
+		   GsBackgroundTile *tile)
+{
+	/* we call the function in idle because the state-change
+	 * notification could be coming from a different thread */
+	g_idle_add (app_state_changed_idle, g_object_ref (tile));
+}
+
+static void
+app_background_tile_css_changed (GsApp *app,
+				 const char *metadata,
+				 GsBackgroundTile *tile)
+{
+	/* just call the app_state_changed function as it will deal with the
+	 * tile CSS already */
+	app_state_changed_idle (g_object_ref (tile));
```

Comment:
I think both this and `app_key_colors_changed()` should end up calling into a function that, after calling `gs_app_get_metadata_item (tile->app, "GnomeSoftware::BackgroundTile-css")` decides whether to set the widget's CSS to the custom background or the key colors one.
That function could be `update_tile_colors_bg()` itself, if it had that additional check, or e.g. another `update_tile_bg()` that does the check.

Reply:
If we do what you say here, then as long as there's a "GnomeSoftware::BackgroundTile-css" metadata, the gradient colors will never be applied, and I think you had mentioned before that if a notify::key-colors occurred, then it should be set independently of what the CSS is. Sorry if I understood it wrong.

Reply:
Correct, that's how I understood the intention of your logic; we either set the custom background, or use the gradient. Your commit message for `GsBackgroundTile` also mentions the same:
```
... has the ability to display either a thumbnail image or a gradient
background generated from an app's icon.
```

So yeah, when either the `GnomeSoftware::BackgroundTile-css` metadata piece or the `key-colors` property change, we should look at both and evaluate which one to use. https://github.com/endlessm/gnome-software/commit/e470c757c072af525ea06520b70d41c7173fc84a has the right logic, and I will comment there with a suggestion for an improvement.